File: frontend/src/features/payments/paymentService.js
import { paymentApi } from "./paymentApi";

/**
 * paymentService chứa các hàm nghiệp vụ phía frontend liên quan đến thanh toán.
 *
 * File này gọi paymentApi để gửi request đến backend,
 * sau đó lấy response.data trả về cho component sử dụng.
 */
export const paymentService = {
    /**
     * Tạo URL thanh toán VNPay cho một đơn hàng.
     * 
     * @param orderId ID của đơn hàng cần thanh toán
     * @returns Dữ liệu backend trả về
     */
    async createVnPayPayment(orderId) {
        const response = await paymentApi.createVnPayPayment(orderId); // Gọi API tạo URL thanh toán VNPay.
        return response.data; // Trả về phần data từ response để component dùng trực tiếp.
    },

    /**
     * Xử lý kết quả thanh toán VNPay trả về sau khi người dùng thanh toán.
     *
     * Hàm này nhận các tham số trên URL redirect từ VNPay,
     * gửi về backend để backend xác thực chữ ký và cập nhật trạng thái thanh toán.
     *
     * @param params Các tham số VNPay trả về trên URL
     *
     * @returns Kết quả xử lý thanh toán từ backend
     */
    async handleVnPayReturn(params) {
        const response = await paymentApi.handleVnPayReturn(params); // Gọi API xử lý kết quả VNPay trả về.
        return response.data; // Trả về message/kết quả từ backend.
    }
};

File: backend/src/main/java/com/cosmetics/ecommerce/service/PaymentService.java
package com.cosmetics.ecommerce.service;

import java.util.Map;

/**
 * Service định nghĩa các nghiệp vụ liên quan đến thanh toán VNPay.
 */
public interface PaymentService {
    /**
     * Tạo URL thanh toán VNPay cho một đơn hàng.
     *
     * Quy trình chính:
     * - Kiểm tra người dùng có quyền thanh toán đơn hàng
     * - Kiểm tra phương thức thanh toán là VNPay
     * - Tạo mã giao dịch txnRef
     * - Build bộ tham số gửi sang VNPay
     * - Ký dữ liệu bằng HMAC SHA512
     * - Trả về URL để frontend redirect người dùng sang VNPay
     *
     * @param userId    ID người dùng đang thực hiện thanh toán
     * @param orderId   ID đơn hàng cần thanh toán
     * @param ipAddress Địa chỉ IP của người dùng thực hiện thanh toán
     * @return URL thanh toán VNPay
     */
    String createVnPayPaymentUrl(Integer userId, Integer orderId, String ipAddress);

    /**
     * Xử lý dữ liệu VNPay trả về sau khi người dùng thanh toán.
     *
     * Quy trình chính:
     * - Nhận toàn bộ params VNPay redirect về
     * - Xác thực chữ ký để đảm bảo dữ liệu không bị chỉnh sửa
     * - Tìm Payment theo mã giao dịch txnRef
     * - Cập nhật trạng thái thanh toán SUCCESS hoặc FAILED
     * - Nếu thanh toán thành công thì cập nhật đơn hàng sang PREPARING
     *
     * @param params Map chứa toàn bộ tham số VNPay trả về
     * @return Thông báo kết quả xử lý thanh toán
     */
    String handleVnPayReturn(Map<String, String> params);
}
